src/2024h2/pubgrub-in-cargo.md

```diff
@@ -0,0 +1,68 @@
+- **Fast**: The resolver can be a slow part of people's workflow. Overall performance must be a high priority and a focus.
+
+## Ownership and other resources
+
+**Owner:** eh2406 will own and lead the effort.
+
+I (eh2406) will be working full time on this effort. I am a member of the Cargo Team and a maintainer of pubgrub-rs.
+
+### Support needed from the project
+
+Integrating the new resolver into Cargo and reaching the shiny future will require extensive collaboration and review from the Cargo Team. However, the next milestones involve independent work exhaustively searching for differences in behavior between the new and old resolvers and fixing them. So only occasional consultation-level conversations will be needed during this proposal.
+
+## Outputs and milestones
+
+### Outputs
+
+Standalone crates for independent components of cargo's resolver. We have already developed https://github.com/pubgrub-rs/pubgrub for solving the core of dependency resolution and https://github.com/pubgrub-rs/semver-pubgrub for doing mathematical set operations on Semver requirements. The shiny future will involve several more crates, although their exact borders have not yet been determined. Eventually we will also be delivering a `-Z pubgrub` for testing the new resolver in cargo itself.
+
+### Milestones
+
+#### For all crate versions on crates.io the two resolvers agree about whether there is a solution.
+
+Build a tool that will look at the index from crates.io and for each version of each crate, make a resolution problem out of resolving the dependencies. This tool will save off an independent test case for each time pubgrub and cargo disagree about whether there *is* a solution. This will not check if the resulting lock files are the same or even compatible, just whether they agree that a lock file *is* possible. Even this crude comparison will find many bugs in how the problem is presented to pubgrub. This is known for sure, because this milestone has already been achieved.
+
+#### For all crate versions on crates.io the two resolvers accept the other one's solution.
+
+The tool from previous milestone will be extended to make sure that the lock file generated by pubgrub can be accepted by cargo's resolver and vice versa. How long will this take? What will this find? No way to know. To quote [FractalFir](https://www.reddit.com/r/rust/comments/1doh929/comment/la9oo2i/) "If I knew where / how many bugs there are, I would have fixed them already. So, providing any concrete timeline is difficult." 
+
+#### For all crate versions on crates.io the performance is acceptable.
+
+There are some crates where pubgrub takes a long time to do resolution, and many more where pubgrub takes longer than cargo's existing resolver. Investigate each of these cases and figure out if performance can be improved either by improvements to the underlying pubgrub algorithm or the way the problem is presented to pubgrub.
```

Comment:
Should we call out prioritization of performance?

For me, the priority is that a successful `cargo metadata --locked` should not be any slower than it is today (I hope I picked the right command for exercising everything).

Resolving on a version req change or `cargo update` is more of an interactive operation and likely could be slower (how much? by a coefficient but likely not an order of magnitude?).

This gives us wiggle room in how we achieve our performance goals.

Reply:
I completely agree on those standards for when we get to merging into cargo. Merging into cargo is not likely in the next six months. More generally, despite my instincts, performance should not be evaluated until correctness has been achieved. So does it need to be called out here? (Either way is fine, I'm just not sure the scope of this document.)